cpp/src/arrow/compute/expression.cc

```diff
@@ -1242,33 +1242,103 @@ struct Inequality {
                             /*insert_implicit_casts=*/false, &exec_context);
   }
 
+  /// Simplify an is_in predicate against this inequality as a guarantee.
+  Result<Expression> SimplifyIsIn(Expression expr) {
+    const auto& guarantee = *this;
+    auto call = expr.call();
+    auto options = checked_pointer_cast<SetLookupOptions>(call->options);
+
+    auto value_set = options->value_set.make_array();
+    if (!value_set) return expr;
+    if (value_set->length() == 0) return literal(false);
+
+    if (!options->sorted_and_deduped) return expr;
```

Comment:
Why without `sorted_and_deduped` the underlying data isn't get checked?

Reply:
If the value set isn't sorted, then we'd have to do a linear scan. For large value sets (the use case that I care about), this cost is prohibitive. Our goal after all is support fast parquet filtering on `is_in`.

However, I could see this being useful for small value sets. Do you think we should add another option, either to `SetLookupOption` or `SimplifyWithGuarantee`, to enable simplifying `is_in` via linear scan? Or can we punt until someone has a use case and actually wants this feature?